File: 0523-continuous-subarray-sum/0523-continuous-subarray-sum.py
class Solution:
    def checkSubarraySum(self, nums, k):
        prefix_sum_map = {}  # remainder of prefix, index
        prefix_sum = 0

        prefix_sum_map[0] = -1
        for i in range(len(nums)):
            prefix_sum += nums[i]
            remainder = prefix_sum % k
            # 현재 나머지가 해시맵에 존재하고, 나머지 값이 이전 값이 아니면
            if remainder in prefix_sum_map and prefix_sum_map[remainder] != i - 1:
                return True
            # 해시맵에 값이 존재하지 않을 때만 추가
            if remainder not in prefix_sum_map:
                prefix_sum_map[remainder] = i

        return False

File: 0523-continuous-subarray-sum/0523-continuous-subarray-sum.java
class Solution {
    /*
        Algorithm & DS: PrefixSum
        Time: O(N), Space: O(N)
    */
    public boolean checkSubarraySum(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>(); // remainder of prefix, index
        int prefixSum = 0;

        map.put(0, -1);
        for(int i = 0; i < nums.length; i++){
            prefixSum += nums[i];
            int remainder = prefixSum % k;
            if(map.containsKey(remainder) && map.get(remainder) != i - 1) return true;
            if(!map.containsKey(remainder)) map.put(remainder, i);
        }

        return false;
    }
}